// TPLinkBulb-api/lib/client.js
'use strict';

const dgram = require('dgram');
const EventEmitter = require('events');

const Bulb = require('./bulb');
const encrypt = require('./utils').encrypt;
const decrypt = require('./utils').decrypt;

const commands = {
  search: '{"system":{"get_sysinfo":{}}}'
};

class Client extends EventEmitter {

  constructor (options) {
    super();
    if (typeof options === 'undefined') options = {};
    this.address = options.address;
    this.port = options.port;
    this.broadcast = options.broadcast || '255.255.255.255';
    this.discoveryInterval = options.discoveryInterval || 30000;
    this.offlineTolerance = options.offlineTolerance || 3;
    this.debug = options.debug || false;

    this.devices = new Map();
    this.discoveryTimer = null;
    this.discoveryPacketSequence = 0;
    this.socket = dgram.createSocket('udp4');

    this.socket.on('error', function (err) {
      this.isSocketBound = false;
      console.error('TP-Link Bulb Client UDP error');
      console.trace(err);
      this.socket.close();
      this.emit('error', err);
    }.bind(this));

    this.socket.on('message', function (msg, rinfo) {
      // Ignore own messages and false formats
      // if (utils.getHostIPs().indexOf(rinfo.address) >= 0 || !Buffer.isBuffer(msg)) {
      //   return
      // }

      const decryptedMsg = decrypt(msg).toString('ascii');

      if (this.debug) {
        console.log('DEBUG - ' + decryptedMsg.toString('ascii') + ' from ' + rinfo.address);
      }

      const jsonMsg = JSON.parse(decryptedMsg);
      const sysinfo = jsonMsg.system.get_sysinfo;
      sysinfo.host = rinfo.address;
      sysinfo.port = rinfo.port;

      if (this.devices.has(sysinfo.deviceId)) {
        const bulb = this.devices.get(sysinfo.deviceId);
        // TODO bulb update here
        bulb.status = 'online';
        bulb.seenOnDiscovery = this.discoveryPacketSequence;
        this.emit('bulb-online', bulb);
      } else {
        const bulb = new Bulb({client: this, deviceId: sysinfo.deviceId, host: rinfo.address, port: rinfo.port, seenOnDiscovery: this.discoveryPacketSequence});

        bulb.name = sysinfo.alias;
        bulb.deviceName = sysinfo.dev_name;
        bulb.model = sysinfo.model;
        bulb.softwareVersion = sysinfo.sw_ver;
        bulb.hardwareVersion = sysinfo.hw_ver;
        bulb.mac = sysinfo.mac;
        bulb.latitude = sysinfo.latitude;
        bulb.longitude = sysinfo.longitude;
        bulb.status = 'online';
        this.devices.set(bulb.deviceId, bulb);
        this.emit('bulb-new', bulb);
      }
    }.bind(this));

    this.socket.bind(this.port, this.address, () => {
      this.socket.setBroadcast(true);
    });
  }

  getBulb (options) {
    return new Bulb(options);
  }

  startDiscovery (devices) {
    this.discoveryTimer = setInterval(() => {
      this.sendDiscovery(devices);
    }, this.discoveryInterval);
    return this.sendDiscovery(devices);
  }

  stopDiscovery () {
    clearInterval(this.discoveryTimer);
    this.discoveryTimer = null;
  }

  sendDiscovery (devices) {
    devices = devices || [];

    this.devices.forEach((device) => {
      if (device.status !== 'offline') {
        var diff = this.discoveryPacketSequence - device.seenOnDiscovery;
        if (diff >= this.offlineTolerance) {
          device.status = 'offline';
          this.emit('bulb-offline', device);
        }
      }
    });

    var msgBuf = encrypt(commands.search);
    this.socket.send(msgBuf, 0, msgBuf.length, 9999, this.broadcast);

    devices.forEach((d) => {
      this.socket.send(msgBuf, 0, msgBuf.length, d.port, d.host);
    });

    if (this.discoveryPacketSequence >= Number.MAX_VALUE) {
      this.discoveryPacketSequence = 0;
    } else {
      this.discoveryPacketSequence += 1;
    }

    return this;
  }

}

module.exports = Client;
